File: gh_raisoni_reports/gh_raisoni_reports/report/cost_centre_and_department_wise_requisition_register/cost_centre_and_department_wise_requisition_register.py


# import frappe


# def execute(filters=None):
#     columns = get_columns()
#     data = get_data(filters or {})
#     return columns, data


# def get_columns():
#     return [
#         {
#             "label":"Material Indent ID",
#             "fieldname":"name",
#             "fieldtype":"Link",
#             "options":"Material Indent",
#             "width":180
#         },
#         {
#             "label":"Transaction Date",
#             "fieldname":"transaction_date",
#             "fieldtype":"Date",
#             "width":130
#         },
#         {
#             "label":"Company",
#             "fieldname":"company",
#             "width":220
#         },
#         {
#             "label":"Department",
#             "fieldname":"department",
#             "width":180
#         },
#         {
#             "label":"Status",
#             "fieldname":"custom_status",
#             "width":180
#         },
#         {
#             "label":"Item Code",
#             "fieldname":"item_code",
#             "width":180
#         },
#         {
#             "label":"Quantity",
#             "fieldname":"qty",
#             "fieldtype":"Float",
#             "width":100
#         }
#     ]


# def get_data(filters):

#     conditions = " WHERE mi.docstatus < 2 "

#     if filters.get("from_date"):
#         conditions += """
#         AND mi.transaction_date >= %(from_date)s
#         """

#     if filters.get("to_date"):
#         conditions += """
#         AND mi.transaction_date <= %(to_date)s
#         """

#     if filters.get("item_code"):
#         conditions += """
#         AND mii.item_code=%(item_code)s
#         """

#     if filters.get("company"):
#         conditions += """
#         AND mi.company=%(company)s
#         """

#     if filters.get("custom_status"):
#         conditions += """
#         AND mi.custom_indent_status=%(custom_status)s
#         """

#     # only department part changed
#     if filters.get("custom_department"):
#         conditions += """
#         AND (
#             CASE
#                 WHEN mi.custom_username LIKE '%% - %%'
#                 THEN TRIM(SUBSTRING_INDEX(mi.custom_username,' - ',-1))
#                 ELSE mi.custom_username
#             END
#         ) = %(custom_department)s
#         """


#     data = frappe.db.sql(f"""
#         SELECT
#             mi.name,
#             mi.transaction_date,
#             mi.company,

#             CASE
#                 WHEN mi.custom_username LIKE '%% - %%'
#                 THEN TRIM(SUBSTRING_INDEX(mi.custom_username,' - ',-1))
#                 ELSE mi.custom_username
#             END as department,

#             mi.custom_indent_status as custom_status,
#             mii.item_code,
#             mii.qty

#         FROM
#             `tabMaterial Indent` mi

#         INNER JOIN
#             `tabMaterial Request Item` mii
#             ON mi.name = mii.parent

#         {conditions}

#         ORDER BY
#             mi.transaction_date DESC

#     """, filters, as_dict=1)

#     return data






# import frappe


# def execute(filters=None):
#     columns = get_columns()
#     data = get_data(filters or {})
#     return columns, data


# def get_columns():
#     return [
#         {
#             "label": "Material Indent ID",
#             "fieldname": "name",
#             "fieldtype": "Link",
#             "options": "Material Indent",
#             "width": 180
#         },
#         {
#             "label": "Transaction Date",
#             "fieldname": "transaction_date",
#             "fieldtype": "Date",
#             "width": 130
#         },
#         {
#             "label": "Company",
#             "fieldname": "company",
#             "width": 220
#         },
#         {
#             "label": "Department",
#             "fieldname": "department",
#             "width": 180
#         },
#         {
#             "label": "Status",
#             "fieldname": "custom_status",
#             "width": 180
#         },
#         {
#             "label": "Item Code",
#             "fieldname": "item_code",
#             "width": 180
#         },
#         {
#             "label": "Quantity",
#             "fieldname": "qty",
#             "fieldtype": "Float",
#             "width": 100
#         }
#     ]


# def get_data(filters):

#     conditions = " WHERE mi.docstatus < 2 "

#     if filters.get("from_date"):
#         conditions += " AND mi.transaction_date >= %(from_date)s"

#     if filters.get("to_date"):
#         conditions += " AND mi.transaction_date <= %(to_date)s"

#     if filters.get("item_code"):
#         conditions += " AND mii.item_code = %(item_code)s"

#     if filters.get("company"):
#         conditions += " AND mi.company = %(company)s"

#     if filters.get("custom_status"):
#         conditions += " AND mi.custom_indent_status = %(custom_status)s"

#     # ✅ FINAL DEPARTMENT FILTER (MIDDLE PART ONLY)
#     if filters.get("custom_department"):
#         conditions += """
#         AND (
#             CASE
#                 WHEN mi.custom_username LIKE '%% - %% - %%'
#                 THEN TRIM(SUBSTRING_INDEX(SUBSTRING_INDEX(mi.custom_username, ' - ', 2), ' - ', -1))
#                 WHEN mi.custom_username LIKE '%% - %%'
#                 THEN TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', 1))
#                 ELSE mi.custom_username
#             END
#         ) = %(custom_department)s
#         """

#     data = frappe.db.sql(f"""
#         SELECT
#             mi.name,
#             mi.transaction_date,
#             mi.company,

#             -- ✅ FINAL DEPARTMENT LOGIC
#             CASE
#                 WHEN mi.custom_username LIKE '%% - %% - %%'
#                 THEN TRIM(SUBSTRING_INDEX(SUBSTRING_INDEX(mi.custom_username, ' - ', 2), ' - ', -1))
#                 WHEN mi.custom_username LIKE '%% - %%'
#                 THEN TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', 1))
#                 ELSE mi.custom_username
#             END as department,

#             mi.custom_indent_status as custom_status,
#             mii.item_code,
#             mii.qty

#         FROM
#             `tabMaterial Indent` mi

#         INNER JOIN
#             `tabMaterial Request Item` mii
#             ON mi.name = mii.parent

#         {conditions}

#         ORDER BY
#             mi.transaction_date DESC

#     """, filters, as_dict=1)

#     return data


# import frappe


# def execute(filters=None):
#     columns = get_columns()
#     data = get_data(filters or {})
#     return columns, data


# def get_columns():
#     return [
#         {
#             "label": "Material Indent ID",
#             "fieldname": "name",
#             "fieldtype": "Link",
#             "options": "Material Indent",
#             "width": 180
#         },
#         {
#             "label": "Transaction Date",
#             "fieldname": "transaction_date",
#             "fieldtype": "Date",
#             "width": 130
#         },
#         {
#             "label": "Company",
#             "fieldname": "company",
#             "width": 220
#         },
#         {
#             "label": "Department",
#             "fieldname": "custom_username",
#             "width": 180
#         },
#         {
#             "label": "Status",
#             "fieldname": "custom_status",
#             "width": 180
#         },
#         {
#             "label": "Item Code",
#             "fieldname": "item_code",
#             "width": 180
#         },
#         {
#             "label": "Quantity",
#             "fieldname": "qty",
#             "fieldtype": "Float",
#             "width": 100
#         }
#     ]


# def get_data(filters):

#     conditions = " WHERE mi.docstatus < 2 "

#     if filters.get("from_date"):
#         conditions += " AND mi.transaction_date >= %(from_date)s"

#     if filters.get("to_date"):
#         conditions += " AND mi.transaction_date <= %(to_date)s"

#     if filters.get("item_code"):
#         conditions += " AND mii.item_code = %(item_code)s"

#     if filters.get("company"):
#         conditions += " AND mi.company = %(company)s"

#     if filters.get("custom_status"):
#         conditions += " AND mi.custom_indent_status = %(custom_status)s"

#     # ✅ SAME FILTER LOGIC (bas correct extraction)
#     if filters.get("custom_department"):
#         conditions += """
#         AND TRIM(SUBSTRING_INDEX(SUBSTRING_INDEX(mi.custom_username, ' - ', 2), ' - ', -1)) = %(custom_department)s
#         """


#     data = frappe.db.sql(f"""
#         SELECT
#             mi.name,
#             mi.transaction_date,
#             mi.company,

#             -- ✅ FINAL FORMAT (Dept - Code)
#             CONCAT(
#                 TRIM(SUBSTRING_INDEX(SUBSTRING_INDEX(mi.custom_username, ' - ', 2), ' - ', -1)),
#                 ' - ',
#                 TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', -1))
#             ) as department,

#             mi.custom_indent_status as custom_status,
#             mii.item_code,
#             mii.qty

#         FROM
#             `tabMaterial Indent` mi

#         INNER JOIN
#             `tabMaterial Request Item` mii
#             ON mi.name = mii.parent

#         {conditions}

#         ORDER BY
#             mi.transaction_date DESC

#     """, filters, as_dict=1)

#     return data


import frappe


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters or {})
    return columns, data


def get_columns():
    return [
        {
            "label": "Material Indent ID",
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Material Indent",
            "width": 180
        },
        {
            "label": "Transaction Date",
            "fieldname": "transaction_date",
            "fieldtype": "Date",
            "width": 130
        },
        {
            "label": "Company",
            "fieldname": "company",
            "width": 220
        },
        # {
        #     "label": "User Name",
        #     "fieldname": "user_name",
        #     "width": 180
        # },
        # {
        #     "label": "Department",
        #     "fieldname": "department",
        #     "width": 180
        # },
        {
            "label": "User & Department",
            "fieldname": "user_and_department",
            "width": 250
        },
        {
            "label": "Status",
            "fieldname": "custom_status",
            "width": 180
        },
        {
            "label": "Item Code",
            "fieldname": "item_code",
            "width": 180
        },
        {
            "label": "Quantity",
            "fieldname": "qty",
            "fieldtype": "Float",
            "width": 100
        }
    ]


def get_data(filters):

    conditions = "WHERE mi.docstatus < 2"

    if filters.get("from_date"):
        conditions += " AND mi.transaction_date >= %(from_date)s"

    if filters.get("to_date"):
        conditions += " AND mi.transaction_date <= %(to_date)s"

    if filters.get("item_code"):
        conditions += " AND mii.item_code = %(item_code)s"

    if filters.get("company"):
        conditions += " AND mi.company = %(company)s"

    if filters.get("custom_status"):
        conditions += " AND mi.custom_indent_status = %(custom_status)s"

    if filters.get("custom_department"):
        conditions += " AND TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', -1)) = %(custom_department)s"

    if filters.get("user_name"):
        filters["user_name"] = f"%{filters['user_name']}%"
        conditions += " AND TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', 1)) LIKE %(user_name)s"

    data = frappe.db.sql(f"""
        SELECT
            mi.name,
            mi.transaction_date,
            mi.company,

            CASE
                WHEN mi.custom_username LIKE '%% - %%'
                THEN TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', 1))
                ELSE mi.custom_username
            END AS user_name,

            CASE
                WHEN mi.custom_username LIKE '%% - %%'
                THEN TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', -1))
                ELSE ''
            END AS department,

            CASE
                WHEN mi.custom_username LIKE '%% - %%'
                THEN CONCAT(
                    TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', 1)),
                    ' - ',
                    TRIM(SUBSTRING_INDEX(mi.custom_username, ' - ', -1))
                )
                ELSE mi.custom_username
            END AS user_and_department,

            mi.custom_indent_status AS custom_status,
            mii.item_code,
            mii.qty

        FROM
            `tabMaterial Indent` mi

        INNER JOIN
            `tabMaterial Request Item` mii
            ON mi.name = mii.parent

        {conditions}

        ORDER BY mi.transaction_date DESC

    """, filters, as_dict=1)

    return data

File: gh_raisoni_reports/gh_raisoni_reports/report/cost_centre_and_department_wise_requisition_register/cost_centre_and_department_wise_requisition_register.js


// frappe.query_reports["Cost Centre and Department wise Requisition Register"] = {
//     "filters": [
//         {
//             "fieldname": "from_date",
//             "label": "From Date",
//             "fieldtype": "Date",
//             "reqd": 1
//         },
//         {
//             "fieldname": "to_date",
//             "label": "To Date",
//             "fieldtype": "Date",
//             "reqd": 1
//         },
//         {
//             "fieldname": "users_department",
//             "label": "Users Department",
//             "fieldtype": "Link",
//             "options": "Department"
//         },
//         {
//             "fieldname": "item_code",
//             "label": "Item Code",
//             "fieldtype": "Link",
//             "options": "Item"
//         }
//     ]
// };


// frappe.query_reports["Cost Centre and Department wise Requisition Register"] = {
//     "filters": [
//         {
//             "fieldname": "from_date",
//             "label": "From Date",
//             "fieldtype": "Date",
//             "reqd": 1
//         },
//         {
//             "fieldname": "to_date",
//             "label": "To Date",
//             "fieldtype": "Date",
//             "reqd": 1
//         },
//         {
//             "fieldname": "custom_username",   // ✅ changed
//             "label": "Username",
//             "fieldtype": "Data"               // ✅ Link nahi, Data
//         },
//         {
//             "fieldname": "item_code",
//             "label": "Item Code",
//             "fieldtype": "Link",
//             "options": "Item"
//         }
//     ]
// };


// frappe.query_reports["Cost Centre and Department wise Requisition Register"] = {
//     "filters": [
//         {
//             "fieldname": "from_date",
//             "label": "From Date",
//             "fieldtype": "Date",
//             "reqd": 1
//         },
//         {
//             "fieldname": "to_date",
//             "label": "To Date",
//             "fieldtype": "Date",
//             "reqd": 1
//         },
//         {
//             "fieldname": "custom_department",   // ✅ Link field
//             "label": "User Department",
//             "fieldtype": "Link",
//             "options": "Department"             // ✅ important
//         },
//         {
//             "fieldname": "item_code",
//             "label": "Item Code",
//             "fieldtype": "Link",
//             "options": "Item"
//         }
//     ]
// };




// frappe.query_reports["Cost Centre and Department wise Requisition Register"] = {
//     filters: [
//         {
//             fieldname:"from_date",
//             label:"From Date",
//             fieldtype:"Date",
//             reqd:1
//         },
//         {
//             fieldname:"to_date",
//             label:"To Date",
//             fieldtype:"Date",
//             reqd:1
//         },
//         {
//             fieldname:"custom_department",
//             label:"User Department",
//             fieldtype:"Link",
//             options:"Department"
//         },
//         {
//             fieldname:"item_code",
//             label:"Item Code",
//             fieldtype:"Link",
//             options:"Item"
//         },
//         {
//             fieldname:"custom_status",
//             label:"Status",
//             fieldtype:"Select",
//             options:"\nOpen\nPurchase Request Raised\nPartially Issued\nClosed"
//         }
//     ]
// };

// frappe.query_reports["Cost Centre and Department wise Requisition Register"] = {
//     filters: [
//         {
//             fieldname:"from_date",
//             label:"From Date",
//             fieldtype:"Date",
//             reqd:1
//         },
//         {
//             fieldname:"to_date",
//             label:"To Date",
//             fieldtype:"Date",
//             reqd:1
//         },
//          {
//             fieldname:"company",
//             label:"Company",
//             fieldtype:"Link",
//             options:"Company"
//         },
//         {
//             fieldname:"custom_department",
//             label:"User Department",
//             fieldtype:"Link",
//             options:"Department"
//         },
       
//         {
//             fieldname:"item_code",
//             label:"Item Code",
//             fieldtype:"Link",
//             options:"Item"
//         },
//         {
//             fieldname:"custom_status",
//             label:"Status",
//             fieldtype:"Select",
//             options:"\nOpen\nPurchase Request Raised\nPartially Issued\nClosed"
//         }
//     ]
// };


frappe.query_reports["Cost Centre and Department wise Requisition Register"] = {

    filters: [
        {
            fieldname: "from_date",
            label: "From Date",
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.month_start()
        },
        {
            fieldname: "to_date",
            label: "To Date",
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.month_end()
        },
        {
            fieldname: "company",
            label: "Company",
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company")
        },
        {
            fieldname: "user_name",
            label: "User Name",
            fieldtype: "Data"
        },
        {
            fieldname: "custom_department",
            label: "Department",
            fieldtype: "Link",
            options: "Department"
        },
        {
            fieldname: "item_code",
            label: "Item Code",
            fieldtype: "Link",
            options: "Item"
        },
        {
            fieldname: "custom_status",
            label: "Status",
            fieldtype: "Select",
            options: "\nOpen\nPurchase Request Raised\nPartially Issued\nClosed"
        }
    ],

    // ✅ Adds Excel & PDF under existing Download button
    onload: function(report) {
        setTimeout(function() {
            report.page.add_inner_button(__("Excel"), function() {
                frappe.query_report.export_report("Excel");
            }, __("Download"));

            report.page.add_inner_button(__("PDF"), function() {
                frappe.query_report.export_report("PDF");
            }, __("Download"));
        }, 1000);
    }
};